docker/server/entrypoint.sh

```diff
@@ -3,164 +3,213 @@
+    DATA_DIR="$(clickhouse extract-from-config --config-file "$CLICKHOUSE_CONFIG" --key=path || true)"
+    TMP_DIR="$(clickhouse extract-from-config --config-file "$CLICKHOUSE_CONFIG" --key=tmp_path || true)"
+    USER_PATH="$(clickhouse extract-from-config --config-file "$CLICKHOUSE_CONFIG" --key=user_files_path || true)"
+    LOG_PATH="$(clickhouse extract-from-config --config-file "$CLICKHOUSE_CONFIG" --key=logger.log || true)"
+    LOG_DIR="$(dirname "$LOG_PATH" || true)"
+    ERROR_LOG_PATH="$(clickhouse extract-from-config --config-file "$CLICKHOUSE_CONFIG" --key=logger.errorlog || true)"
+    ERROR_LOG_DIR="$(dirname "$ERROR_LOG_PATH" || true)"
+    FORMAT_SCHEMA_PATH="$(clickhouse extract-from-config --config-file "$CLICKHOUSE_CONFIG" --key=format_schema_path || true)"
+
+    declare -g CLICKHOUSE_USER CLICKHOUSE_PASSWORD CLICKHOUSE_DB CLICKHOUSE_ACCESS_MANAGEMENT
+    CLICKHOUSE_USER="${CLICKHOUSE_USER:-default}"
+    CLICKHOUSE_PASSWORD="${CLICKHOUSE_PASSWORD:-}"
+    CLICKHOUSE_DB="${CLICKHOUSE_DB:-}"
+    CLICKHOUSE_ACCESS_MANAGEMENT="${CLICKHOUSE_DEFAULT_ACCESS_MANAGEMENT:-0}"
+    
+    declare -g DATABASE_ALREADY_EXISTS
+	if [ -d "$DATA_DIR/data" ]; then
+		DATABASE_ALREADY_EXISTS='true'
+	fi
+}
+
+setup_dirs() {
+    for dir in "$DATA_DIR" \
+    "$ERROR_LOG_DIR" \
+    "$LOG_DIR" \
+    "$TMP_DIR" \
+    "$USER_PATH" \
+    "$FORMAT_SCHEMA_PATH"
+    do
+        # check if variable not empty
+        [ -z "$dir" ] && continue
+        # ensure directories exist
+        if ! mkdir -p "$dir"; then
+            echo "Couldn't create necessary directory: $dir"
+            exit 1
+        fi
 
-# if clickhouse user is defined - create it (user "default" already exists out of box)
-if [ -n "$CLICKHOUSE_USER" ] && [ "$CLICKHOUSE_USER" != "default" ] || [ -n "$CLICKHOUSE_PASSWORD" ]; then
+        if [ "$DO_CHOWN" = "1" ]; then
+            # ensure proper directories permissions
+            chown -R "$USER:$GROUP" "$dir"
+        elif ! $gosu test -d "$dir" -a -w "$dir" -a -r "$dir"; then
+            echo "Necessary directory '$dir' isn't accessible by user with id '$USER'"
+            exit 1
+        fi
+    done
+}
+
+create_clickhouse_user() {
+    # if clickhouse user is defined - create it (user "default" already exists out of box)
+    if [ -n "$CLICKHOUSE_USER" ] && [ "$CLICKHOUSE_USER" != "default" ] || [ -n "$CLICKHOUSE_PASSWORD" ]; then
     echo "$0: create new user '$CLICKHOUSE_USER' instead 'default'"
     cat <<EOT > /etc/clickhouse-server/users.d/default-user.xml
     <yandex>
-      <!-- Docs: <https://clickhouse.tech/docs/en/operations/settings/settings_users/> -->
-      <users>
+    <!-- Docs: <https://clickhouse.tech/docs/en/operations/settings/settings_users/> -->
+    <users>
         <!-- Remove default user -->
         <default remove="remove">
         </default>
 
         <${CLICKHOUSE_USER}>
-          <profile>default</profile>
-          <networks>
+        <profile>default</profile>
```

Comment:
Why flattening xml?

Reply:
Maybe i'm wrong but for me is equal to the original entrypoint.sh

Reply:
whitespaces at the begining of line